### PR TITLE
Honor PEP 621 `requires-python` setting.

### PR DESCRIPTION
**Pull request by sarayourfriend**
Fixes https://github.com/beeware/briefcase/issues/2016

While I believe this changeset addresses the issue as described, I have some lingering questions about the approach I took. Namely, the Linux system command mixins already include some Python interpreter version validation to ensure the running interpreter is compatible with Briefcase, and I haven't found a good way to integrate that into this change. I wondered for a bit whether a basic definition of `LinuxSystemMostlyPassiveMixin::verify_python` should be hoisted to the `BaseCommand`, and then `super()` called from those Linux system commands instead of creating an entirely new method on the `BaseCommand`. I chose against trying to do that because it seemed the concerns were getting mixed a bit, and I wanted to see what the maintainers thought before making any bigger changes.

If I were going to fully integrate this new bit of interpreter version validation with the existing bits, I think I'd like to rename `verify_python` to `verify_interpreter_briefcase_compatibility` (or something along those lines). Then name the `BaseCommand` method just `verify_python` and call to `verify_interpreter_briefcase_compatibility` from an override of that base command method in the mixin.

Speaking generally, it would probably be a good idea to rename `verify_python` in that mixin anyway, to at least nominally clarify the difference between its responsibility and that of the base command's own Python interpreter verification. But, like I said, I wasn't too confident in any of the approaches I could come up with, and wanted to avoid creating a bigger changeset than absolutely necessary. That being said, I'm happy to keep working on this to create a more cohesive solution, just let me know :slightly_smiling_face:.

Also: thank you for the excellent contributor documentation. It was easy to get started contributing to BeeWare! And a pleasant codebase to navigate around too.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Initial quick review looks really good - will do a full review when CI is passing.

From the look of it, the issue is that `config.py` hasn't yet acquired the `from __future__ import annotations` that enables PEP 563 deferred annotation handling. This has been added to *most* files in Briefcase;l but for some reason, `config.py` hasn't been updated yet.

**Comment by sarayourfriend** [comment]
Aha! Thanks, I previously pushed a commit to just remove that annotation, but have replaced it with the future import :+1:

File: tests/commands/base/test_verify_requires_python.py
Comment (sarayourfriend) [inline tests/commands/base/test_verify_requires_python.py:35]:
```suggestion
    """Validation fails if requires-python specifies a version greater than the running interpreter."""
```

**Comment by sarayourfriend** [comment]
Okay, seems like the CI is good aside from the problem fixed in https://github.com/beeware/briefcase/pull/2028 :tada:

File: tests/commands/base/test_verify_requires_python.py
Comment (freakboy3742) [inline tests/commands/base/test_verify_requires_python.py:28]: This is a good test; but it would be better if it were extended with `pytest.mark.parameterize` to verify a range of specifiers that meet requirements (>, >=, <, <=, ==), rather than a single "lower than" test.

File: tests/commands/base/test_verify_requires_python.py
Comment (freakboy3742) [inline tests/commands/base/test_verify_requires_python.py:38]: As with the "met" test, this would be better if parameterised to cover more possible values.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
This looks really good! A couple of small suggestions inline; we'll also need to wait until #2028 lands to confirm we're good to merge.

Regarding your comments about the Linux version checks - those are slightly different, because they are quite often verifying the version of Python in a Docker container that Briefcase is managing, rather than Briefcase itself. To that end, they're a slightly different verification to what is being validated here.

However, I think your instinct about renaming those validations from `verify_python` to something else is sound; `verify_interpreter_briefcase_compatibility` is a little more  verbose than I'd prefer. Something like `verify_system_python` would be a little more palatable.

**Comment by freakboy3742** [comment]
FYI - #2028 has now landed, so if you merge with main, CI should now pass.

File: src/briefcase/commands/base.py
Comment (sarayourfriend) [reply src/briefcase/commands/base.py:682]: TIL, thanks!

File: tests/commands/base/test_verify_requires_python.py
Comment (sarayourfriend) [reply tests/commands/base/test_verify_requires_python.py]: Good idea, done in https://github.com/beeware/briefcase/pull/2029/commits/dde2dafcc01646aceab4475d4e4718d9c539fa76

File: tests/commands/base/test_verify_requires_python.py
Comment (sarayourfriend) [reply tests/commands/base/test_verify_requires_python.py]: :+1: Done in https://github.com/beeware/briefcase/pull/2029/commits/dde2dafcc01646aceab4475d4e4718d9c539fa76

File: src/briefcase/commands/base.py
Comment (sarayourfriend) [reply src/briefcase/commands/base.py]: Done in https://github.com/beeware/briefcase/pull/2029/commits/4aa29d535fc7a4a2c51810879704c7e9d612c465

**Comment by sarayourfriend** [comment]
You may be partial to `verify_system_python` because it already exists :sweat_smile: 

https://github.com/beeware/briefcase/blob/63f726094058d43d7e61545ed91c8ac02b1293b4/src/briefcase/platforms/linux/system.py#L447

After looking at the usage of the existing `verify_python` method a bit, it seems to _only_ be used for validating the Docker container's interpreter. The method's docstring claims this, and it is called only in this branch guarded by `if self.use_docker`:

https://github.com/beeware/briefcase/blob/63f726094058d43d7e61545ed91c8ac02b1293b4/src/briefcase/platforms/linux/system.py#L578-L594

I've renamed it to `verify_docker_python`, which matches the verbosity of the other methods, and (hopefully) accurately reflects the method's scope. Let me know what you think, happy to use any name you'd like or leave this change out of this PR altogether.

**Review by freakboy3742 — APPROVED** [review]
I've made a small tweak to the parameterized cases to (hopefully) make them a little more clear, and be a little less reliant on the specific Python version - but otherwise, this looks great - Thanks for the contribution! 

As luck has it, you've wrapped this up just before I was about to push a release, so this feature should be in the wild in a couple of hours as part of Briefcase 0.3.20.

(On an unrelated note - I notice from your bio that you're (a) looking for work, and (b) based in naarm; if you're not already planning to be at [PyCon AU 2024](https://2024.pycon.org.au), that should be some companies looking to hire there - and, for bonus points, you'll be able to claim your [BeeWare challenge coin](https://beeware.org/contributing/challenge-coins/), because I'll be there :-))

**Comment by sarayourfriend** [comment]
> (On an unrelated note - I notice from your bio that you're (a) looking for work, and (b) based in naarm; if you're not already planning to be at [PyCon AU 2024](https://2024.pycon.org.au/), that should be some companies looking to hire there - and, for bonus points, you'll be able to claim your [BeeWare challenge coin](https://beeware.org/contributing/challenge-coins/), because I'll be there :-))

Thanks! The timing of this contribution may or may not be related to my presence at PyCon AU next month and having seen you speak at KiwiPycon last month :slightly_smiling_face:. I very much enjoyed your talk in Wellington, and suddenly have a lot of time on my hands to explore contributions to projects other than Openverse, so thought I'd take your advice from the Q&A at that talk and look at BeeWare.

I'll see you at PyCon, and around GitHub :grin:

**Pull request merged**
